### PR TITLE
improve dws_flex ux

### PR DESCRIPTION
**Pull request by abbas1902**
### Submission Checklist

NOTE: Community submissions can take up to 2 weeks to be reviewed.

Please take the following actions before submitting this pull request.

* Fork your PR branch from the Toolkit "develop" branch (not main)
* Test all changes with pre-commit in a local branch [#](https://goo.gle/hpc-toolkit#development)
* Confirm that "make tests" passes all tests
* Add or modify unit tests to cover code changes
* Ensure that unit test coverage remains above 80%
* Update all applicable documentation
* Follow Cluster Toolkit Contribution guidelines [#](https://goo.gle/hpc-toolkit-contributing)

### CODE REVIEW

File: community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/slurm_files/scripts/resume.py [hunk 1]
```diff
@@ -101,8 +101,16 @@ def instance_properties(nodeset, model, placement_group, labels=None):
     # Override with properties explicit specified in the nodeset
     props.update(nodeset.get("instance_properties") or {})
     
+    if (nodeset.dws_flex.enabled):
```
Comment (mr0re1) [inline community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/slurm_files/scripts/resume.py:104]: It should be called before line 101. `nodeset.instance_properties` should override whatever our scripts came up with.

File: community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/slurm_files/scripts/resume.py [hunk 2]
```diff
@@ -101,8 +101,16 @@ def instance_properties(nodeset, model, placement_group, labels=None):
...
+        update_props_dws(props,nodeset.dws_flex)
+
     return props
 
+def update_props_dws(props,dws_flex):
```
Comment (mr0re1) [inline community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/slurm_files/scripts/resume.py:109]: Please add type annotation, simple one will do:

`def update_props_dws(props: object,dws_flex: object): -> None`

File: community/modules/compute/schedmd-slurm-gcp-v6-nodeset/variables.tf [hunk 3]
```diff
@@ -512,3 +512,29 @@ variable "enable_maintenance_reservation" {
...
+  If set and `enabled = true`, will utilize the DWS Flex Start to provision nodes.
+  See: https://cloud.google.com/blog/products/compute/introducing-dynamic-workload-scheduler
+  Options:
+  - enable: Enable DWS Flex Start
+  - max_run_duration: Maximum duration in seconds for the job to run, should not exceed 1,209,600 (2 weeks).
+  - use_job_duration: Use the job duration to determine the max_run_duration, if job duration is not set, max_run_duration will be used.
+  
+ Limitations:
+  - CAN NOT be used with reservations;
+  - CAN NOT be used with placement groups (TODO: revisit);
```
Comment (mr0re1) [inline community/modules/compute/schedmd-slurm-gcp-v6-nodeset/variables.tf:527]: Let's keep this limitation (remove constraint)
Also we need to enforce them, see `community/modules/compute/schedmd-slurm-gcp-v6-nodeset/outputs.tf` for example on how we do it.

File: community/modules/compute/schedmd-slurm-gcp-v6-nodeset/main.tf [hunk 4]
```diff
@@ -52,12 +52,18 @@ locals {
     content  = var.startup_script
   }]
 
+  dws_flex = {
+    enabled = var.dws_flex.enabled
+    max_run_duration = tostring(var.dws_flex.max_run_duration)
```
Comment (mr0re1) [inline community/modules/compute/schedmd-slurm-gcp-v6-nodeset/main.tf:57]: why `tostring` ?

File: community/modules/scheduler/schedmd-slurm-gcp-v6-controller/variables.tf [hunk 5]
```diff
@@ -212,6 +212,10 @@ variable "nodeset" {
       count = number
       type  = string
     }))
+    dws_flex = optional(object({
```
Comment (mr0re1) [inline community/modules/scheduler/schedmd-slurm-gcp-v6-controller/variables.tf:215]: Let's not make it `optional` here. Controller should expected fully populated `nodeset`-entity. It's nodeset-module job to populate while providing user conveniences.

File: community/modules/compute/schedmd-slurm-gcp-v6-nodeset/variables.tf [hunk 6]
```diff
@@ -512,3 +512,29 @@ variable "enable_maintenance_reservation" {
...
+
+ EOD
+
+ type = object({
+   enabled = bool
```
Comment (mr0re1) [inline community/modules/compute/schedmd-slurm-gcp-v6-nodeset/variables.tf:532]: `optional(default, true)`

File: community/modules/compute/schedmd-slurm-gcp-v6-nodeset/variables.tf [hunk 7]
```diff
@@ -512,3 +512,29 @@ variable "enable_maintenance_reservation" {
...
+   max_run_duration = number
```
Comment (mr0re1) [inline community/modules/compute/schedmd-slurm-gcp-v6-nodeset/variables.tf:533]: `optional(number, 2weeks)`

File: community/modules/compute/schedmd-slurm-gcp-v6-nodeset/variables.tf (hunk 6 above)
Comment (mr0re1) [reply community/modules/compute/schedmd-slurm-gcp-v6-nodeset/variables.tf:532]: We can discuss it offline, if it doesn't makes sense / you have any concerns.

File: community/modules/compute/schedmd-slurm-gcp-v6-nodeset/main.tf (hunk 4 above)
Comment (abbas1902) [reply community/modules/compute/schedmd-slurm-gcp-v6-nodeset/main.tf]: I think I had some odd error when using it as a number, in any case the error was unrelated so no need to convert the type.

File: community/modules/compute/schedmd-slurm-gcp-v6-nodeset/variables.tf (hunk 6 above)
Comment (abbas1902) [reply community/modules/compute/schedmd-slurm-gcp-v6-nodeset/variables.tf]: Added it, had a question about it but will take it offline

File: community/modules/compute/schedmd-slurm-gcp-v6-nodeset/outputs.tf [hunk 8]
```diff
@@ -44,4 +44,13 @@ output "nodeset" {
...
     error_message = "Cannot use placement with static and auto-scaling nodes in the same node set."
   }
+  precondition {
+    condition     = var.reservation_name == "" || !var.dws_flex.enabled
+    error_message = "Cannot use reservations with DWS Flex."
+  }
+
+  precondition {
+    condition     = !var.enable_placement || !var.dws_flex.enabled
+    error_message = "Cannot use DWS Flex with an existing placement policy."
```
Comment (mr0re1) [inline community/modules/compute/schedmd-slurm-gcp-v6-nodeset/outputs.tf:54]: `enable_placement` creates ad-hoc placement policies. Please change wording "with an existing placement policy" -> "with `enable_placement `"

File: docs/slurm-dws-flex.md [hunk 9]
```diff
@@ -13,25 +13,19 @@ With Dynamic Workload Scheduler in Flex Start mode, you submit a GPU capacity re
...
     settings:
-      instance_properties:
-        reservationAffinity:
-          consumeReservationType: NO_RESERVATION
-        scheduling:
-          maxRunDuration: { seconds: $(2 * 60 * 60) } # 2 hours
-          onHostMaintenance: TERMINATE
-          instanceTerminationAction: DELETE
+       dws_flex:
+          enabled: true
```
Comment (mr0re1) [inline docs/slurm-dws-flex.md:24]: nit. We can omit `enabled` to appear more slick ))

**Comment by mr0re1** [comment]
Nice, LGTM

**Review by mr0re1 — APPROVED** [review]

File: docs/slurm-dws-flex.md (hunk 9 above)
Comment (abbas1902) [reply docs/slurm-dws-flex.md]: Sure, I also added `enable_placement: false` since that will be needed for nodesets using dws_flex

**Pull request merged**